### PR TITLE
Mimics

### PR DESCRIPTION
**Pull request by Oricana-16**
# About The Pull Request

Mimics-
They can disguise as items, and can latch onto people. When latched on, the person they are latched Onto can't attack them back. After they kill someone or find a corpse, they can absorb it and then replicate.

## Why It's Good For The Game

A fun new antag idea that'll scare people

## Changelog

:cl:
add: Adds mimics to the game, a stealth-based antagonist that can disguise and replicate. 
/:cl:

### CODE REVIEW

**Comment by nednaZ** [comment]
If you dig into what this is, don't spoil it for others.

**Comment by Oricana-16** [comment]
Since a lot of people know about them now, I assume its fine to undisguise this PR

**Comment by nednaZ** [comment]
Feel free to have it be open with a proper changelog now.
Also, I would like them added as a rare antagonist in Dynamic if possible rather than it being 100% a random event.

File: monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+//           MIMICS (GHOST)                 //
+//                                          //
+//////////////////////////////////////////////
+
+/datum/dynamic_ruleset/midround/from_ghosts/mimic
+	name = "Mimic"
+	antag_datum = /datum/antagonist/mimic
+	antag_flag = "Mimic"
+	antag_flag_override = ROLE_ALIEN
+	enemy_roles = list("AI", "Security Officer", "Detective", "Warden", "Head of Security", "Captain")
```
Comment (nednaZ) [inline monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:12]: Remove AI from this, they can't fight mimics directly

File: monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	required_enemies = list(4,3,2,2,1,1,1,1,0,0)
```
Comment (nednaZ) [inline monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:13]: Change it to be 3 down the line

File: monkestation/code/modules/antagonists/mimic/mimic.dm [hunk 3]
```diff
@@ -0,0 +1,561 @@
+#define MIMIC_HEALTH_FLEE_AMOUNT 30
+#define MIMIC_JITTER_CHANCE 0.8
+#define MIMIC_DISGUISE_COOLDOWN 5 SECONDS
```
Comment (nednaZ) [inline monkestation/code/modules/antagonists/mimic/mimic.dm:3]: Don't forget to #undef them too

File: monkestation/code/modules/antagonists/mimic/mimic.dm [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+	stop_automated_movement = 1
+	status_flags = CANPUSH
+	stat_attack = UNCONSCIOUS
+	mob_size = MOB_SIZE_SMALL
+	pass_flags = PASSTABLE | PASSMOB
+	unsuitable_atmos_damage = 0 //They won't die in Space!
+	minbodytemp = TCMB
+	maxbodytemp = T0C + 40
+	maxHealth = 90
+	health = 90
```
Comment (nednaZ) [inline monkestation/code/modules/antagonists/mimic/mimic.dm:28]: Boost them to 120, I think a 25% increase will be fair based off the playtesting we have done.

File: monkestation/code/modules/antagonists/mimic/mimic.dm [hunk 5]
```diff
@@ -0,0 +1,561 @@
...
+	. = ..()
+
+/mob/living/simple_animal/hostile/alien_mimic/proc/allowed(atom/movable/target_item)
+	return isitem(target_item) && !istype(target_item, /obj/item/radio/intercom)
+
+/mob/living/simple_animal/hostile/alien_mimic/proc/is_table(atom/possible_table)
+	return istype(possible_table, /obj/structure/table) || istype(possible_table, /obj/structure/rack)
+
+/mob/living/simple_animal/hostile/alien_mimic/proc/replication_cost()
+	return 1 + round((mimic_count / 2))
```
Comment (nednaZ) [inline monkestation/code/modules/antagonists/mimic/mimic.dm:110]: Unneeded proc, just put it right into where it needs to be OR create a macro for it.

File: monkestation/code/modules/antagonists/mimic/mimic.dm [hunk 6]
```diff
@@ -0,0 +1,561 @@
...
+
+	cut_overlays()
+	set_varspeed(move_to_delay)
+	med_hud_set_health()
+	med_hud_set_status() //we are not an object
+
+/mob/living/simple_animal/hostile/alien_mimic/Initialize(mapload)
+	if(!hivemind_name)
+		//1% chance for some silly names
+		hivemind_name = rand(99) ? "Mimic [rand(1,999)]" : pick("John","Not-A-Mimic","Goop Spider","Syndicate Infiltrator","Mimic Hater","Nar'sie Enthusiast")
```
Comment (nednaZ) [inline monkestation/code/modules/antagonists/mimic/mimic.dm:271]: That should be a prob(99) rather than rand(99)

**Review by nednaZ — CHANGES_REQUESTED** [review]
Some changes before a final merge

File: monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm (hunk 2 above)
Comment (Oricana-16) [reply monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:13]: Im assuming you mean like replace more with 3's, if not lmk

File: monkestation/code/modules/antagonists/mimic/mimic.dm [hunk 7]
```diff
@@ -0,0 +1,564 @@
...
+		var/mob/recipient = player
+		if(ismimic(recipient))
+			to_chat(recipient, my_message)
+		else if(recipient in GLOB.dead_mob_list)
+			var/link = FOLLOW_LINK(recipient, user)
+			to_chat(recipient, "[link] [my_message]")
+
+	user.log_talk(message, LOG_SAY, tag="mimic hivemind")
+
+#undef MIMIC_HEALTH_FLEE_AMOUNT 30
```
Comment (Zonespace27) [inline monkestation/code/modules/antagonists/mimic/mimic.dm:562]: undef is done incorrectly

Comment (Oricana-16) [reply monkestation/code/modules/antagonists/mimic/mimic.dm:562]: guh im silly

File: monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm (hunk 2 above)
Comment (nednaZ) [reply monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: list(3,3,3,3,3,3,3,3,3,3)

File: monkestation/code/modules/antagonists/mimic/mimic.dm [hunk 8]
```diff
@@ -0,0 +1,569 @@
...
+	real_name = "mimic"
+	desc = "A morphing mass of black gooey tendrils."
+	speak_emote = list("warbles")
+	emote_hear = list("warbles")
+	faction = list("aliens") //don't wanna have them attack eachother
+	icon = 'monkestation/icons/mob/animal.dmi'
+	icon_state = "mimic"
+	icon_living = "mimic"
+	icon_dead = "mimic_dead"
+	move_to_delay = 5
```
Comment (nednaZ) [inline monkestation/code/modules/antagonists/mimic/mimic.dm:18]: Seconds defines
```suggestion
	move_to_delay = 0.5 SECONDS
```

File: monkestation/code/modules/antagonists/mimic/mimic.dm [hunk 9]
```diff
@@ -0,0 +1,569 @@
...
+			latch(victim)
+			return
+	if(!ismob(target)) //If you're attacking something or other
+		if(disguised)
+			restore()
+		return ..()
+	if(buckled && target == buckled) //If you're buckled to them, or attacking a non-human
+		return ..()
+	if(iscyborg(target) || isAI(target)) //stinky sillicons with their no mounting rules
+		return ..()
```
Comment (nednaZ) [inline monkestation/code/modules/antagonists/mimic/mimic.dm:400]: This fires too late, the do_mob still attempts to run before actually attacking.

**Review by nednaZ — CHANGES_REQUESTED** [review]
Probably the last needed changes

File: monkestation/code/modules/antagonists/mimic/mimic_antag.dm [hunk 10]
```diff
@@ -0,0 +1,11 @@
+/datum/antagonist/mimic
```
Comment (nednaZ) [inline monkestation/code/modules/antagonists/mimic/mimic_antag.dm:1]: Add show_to_ghosts 1 so observers can easily watch

File: monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+//                                          //
+//////////////////////////////////////////////
+
+/datum/dynamic_ruleset/midround/from_ghosts/mimic
+	name = "Mimic"
+	antag_datum = /datum/antagonist/mimic
+	antag_flag = "Mimic"
+	antag_flag_override = ROLE_ALIEN
+	enemy_roles = list("Security Officer", "Detective", "Warden", "Head of Security", "Captain")
+	required_enemies = list(4,3,3,3,2,2,1,1,1,0)
```
Comment (nednaZ) [inline monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:13]:
```suggestion
	required_enemies = list(2,2,2,2,2,2,2,1,1,0)
```

File: monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+	required_candidates = 2
+	weight = 2
```
Comment (nednaZ) [inline monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:15]:
```suggestion
	weight = 3
```

File: monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+	cost = 15
+	minimum_players = 30
+	requirements = list(101,101,101,60,50,40,20,15,10,10)
```
Comment (nednaZ) [inline monkestation/code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:18]:
```suggestion
	requirements = list(101,101,101,30,30,30,30,30,10,10)
```

**Review by nednaZ — APPROVED** [review]
Final changes, I'm just going to commit them myself.
After plenty of test play and balancing, these are ready to go.

**Pull request merged**